docker/utils/build.py

```diff
@@ -23,127 +27,90 @@ def exclude_paths(root, patterns, dockerfile=None):
 
     All paths returned are relative to the root.
     """
+
     if dockerfile is None:
         dockerfile = 'Dockerfile'
 
-    patterns = [p.lstrip('/') for p in patterns]
-    exceptions = [p for p in patterns if p.startswith('!')]
-
-    include_patterns = [p[1:] for p in exceptions]
-    include_patterns += [dockerfile, '.dockerignore']
-
-    exclude_patterns = list(set(patterns) - set(exceptions))
-
-    paths = get_paths(root, exclude_patterns, include_patterns,
-                      has_exceptions=len(exceptions) > 0)
-
-    return set(paths).union(
-        # If the Dockerfile is in a subdirectory that is excluded, get_paths
-        # will not descend into it and the file will be skipped. This ensures
-        # it doesn't happen.
-        set([dockerfile.replace('/', os.path.sep)])
-        if os.path.exists(os.path.join(root, dockerfile)) else set()
-    )
-
-
-def should_include(path, exclude_patterns, include_patterns, root):
-    """
-    Given a path, a list of exclude patterns, and a list of inclusion patterns:
-
-    1. Returns True if the path doesn't match any exclusion pattern
-    2. Returns False if the path matches an exclusion pattern and doesn't match
-       an inclusion pattern
-    3. Returns true if the path matches an exclusion pattern and matches an
-       inclusion pattern
-    """
-    for pattern in exclude_patterns:
-        if match_path(path, pattern):
-            for pattern in include_patterns:
-                if match_path(path, pattern):
-                    return True
-                if os.path.isabs(pattern) and match_path(
-                        os.path.join(root, path), pattern):
-                    return True
-            return False
-    return True
-
-
-def should_check_directory(directory_path, exclude_patterns, include_patterns,
-                           root):
+    def normalize(p):
+        # Leading and trailing slashes are not relevant. Yes,
+        # "foo.py/" must exclude the "foo.py" regular file. "."
+        # components are not relevant either, even if the whole
+        # pattern is only ".", as the Docker reference states: "For
+        # historical reasons, the pattern . is ignored."
+        splitted = [pt for pt in re.split(_SEP, p) if pt and pt != '.']
+        # ".." component must be cleared with the potential previous
+        # component, regardless of whether it exists: "A preprocessing
+        # step [...]  eliminates . and .. elements using Go's
+        # filepath.".
+        i = 0
+        while i < len(splitted):
+            if splitted[i] == '..':
+                del splitted[i]
+                if i > 0:
```

Comment:
Does this mean that `../xxx` becomes `xxx`? If so, does it match the behavior of the engine?

Reply:
More generally, having some tests for `..` components would be nice

Reply:
The official documentation about this is rather elusive. It does however stipulate that absolute paths are equivalent to relative paths, hence `/../foo` should be equivalent to `../foo`. It also stipulates that paths are run through Go's `filepath.Clean`, which explicitely makes this equivalence: ```Eliminate .. elements that begin a rooted path:
   that is, replace "/.." by "/" at the beginning of a path
```

From this, I would argue that yes, the specification prompts for ../foo to be equivalent to foo.

But as you smartly suggested, I tested with Docker `17.09.1-ce`, and `../foo` does not ignore a `foo` file at the root - I suppose it won't ignore anything ever.

However, brace yourself, if you use `/../foo`, it *does* ignore a `foo` file at the root, which violates the statements that absolute and relative paths are equivalent. Long story short: it's a mess.

It might be more reasonable to mimic the current engine behavior, although not following the spec could seem worse. The spec would probably need some updating if we choose to mimic the engine.

Reply:
Ok - it probably doesn't see a lot of use anyway. I say we keep it as is and we can always revisit it later if it's an issue.

Reply:
I agree that even any `..` in a `.dockerignore` file doesn't make much sense to me. Especially since `anything/..` will be dropped, even if `anything` does not exist as a directory, so I really don't see the point. IMO we should even deprecate and then reject them.